### DISCUSSION TITLE
How to build OpenSsl 3.0 to enable all ciphersuites

### DISCUSSION THREAD

**Discussion by Tamec1** [openssl/openssl#22092] [Q&A] [answered]
I built OpenSSL 3.0.8 with the -enable-weak-ssl-ciphers flag 
However, if I try to find ciphersuites via rfc name I get nullptrs for some suites.

I am searching e.g. like this (simplificated):
`
auto name = OPENSSL_cipher_name("TLS_RSA_WITH_DES_CBC_SHA");
`
Debugging this code revealed that this calls 
`
ssl3_get_cipher_by_std_name(const char *stdname)
`
Therein, among other things, the ssl3_ciphers table is searched for the stdname in it,
however there are only 197 entries at all in this struct (with the weak-ssl-ciphers enabled by the build flag), 
I would expect more. 

We are working on a test tool (reason why we need to support older and deprecated suites as well) 
and previously used the mbed lib here is a full list of ciphers we previously supported with mbed
but which I now cant use with my openssl build:

TLS_RSA_WITH_DES_CBC_SHA
TLS_DHE_RSA_WITH_DES_CBC_SHA
TLS_ECDH_ECDSA_WITH_NULL_SHA
TLS_ECDH_ECDSA_WITH_RC4_128_SHA
TLS_ECDH_ECDSA_WITH_3DES_EDE_CBC_SHA
TLS_ECDH_ECDSA_WITH_AES_128_CBC_SHA
TLS_ECDH_ECDSA_WITH_AES_256_CBC_SHA
TLS_ECDH_RSA_WITH_NULL_SHA
TLS_ECDH_RSA_WITH_RC4_128_SHA
TLS_ECDH_RSA_WITH_3DES_EDE_CBC_SHA
TLS_ECDH_RSA_WITH_AES_128_CBC_SHA
TLS_ECDH_RSA_WITH_AES_256_CBC_SHA
TLS_ECDH_ECDSA_WITH_AES_128_CBC_SHA256
TLS_ECDH_ECDSA_WITH_AES_256_CBC_SHA384
TLS_ECDH_RSA_WITH_AES_128_CBC_SHA256
TLS_ECDH_RSA_WITH_AES_256_CBC_SHA384
TLS_ECDH_ECDSA_WITH_AES_128_GCM_SHA256
TLS_ECDH_ECDSA_WITH_AES_256_GCM_SHA384
TLS_ECDH_RSA_WITH_AES_128_GCM_SHA256
TLS_ECDH_RSA_WITH_AES_256_GCM_SHA384
TLS_RSA_WITH_ARIA_128_CBC_SHA256
TLS_RSA_WITH_ARIA_256_CBC_SHA384
TLS_DHE_RSA_WITH_ARIA_128_CBC_SHA256
TLS_DHE_RSA_WITH_ARIA_256_CBC_SHA384
TLS_ECDHE_ECDSA_WITH_ARIA_128_CBC_SHA256
TLS_ECDHE_ECDSA_WITH_ARIA_256_CBC_SHA384
TLS_ECDH_ECDSA_WITH_ARIA_128_CBC_SHA256
TLS_ECDH_ECDSA_WITH_ARIA_256_CBC_SHA384
TLS_ECDHE_RSA_WITH_ARIA_128_CBC_SHA256
TLS_ECDHE_RSA_WITH_ARIA_256_CBC_SHA384
TLS_ECDH_RSA_WITH_ARIA_128_CBC_SHA256
TLS_ECDH_RSA_WITH_ARIA_256_CBC_SHA384
TLS_ECDH_ECDSA_WITH_ARIA_128_GCM_SHA256
TLS_ECDH_ECDSA_WITH_ARIA_256_GCM_SHA384
TLS_ECDH_RSA_WITH_ARIA_128_GCM_SHA256
TLS_ECDH_RSA_WITH_ARIA_256_GCM_SHA384
TLS_PSK_WITH_ARIA_128_CBC_SHA256
TLS_PSK_WITH_ARIA_256_CBC_SHA384
TLS_DHE_PSK_WITH_ARIA_128_CBC_SHA256
TLS_DHE_PSK_WITH_ARIA_256_CBC_SHA384
TLS_RSA_PSK_WITH_ARIA_128_CBC_SHA256
TLS_RSA_PSK_WITH_ARIA_256_CBC_SHA384
TLS_ECDHE_PSK_WITH_ARIA_128_CBC_SHA256
TLS_ECDHE_PSK_WITH_ARIA_256_CBC_SHA384
TLS_ECDH_ECDSA_WITH_CAMELLIA_128_CBC_SHA256
TLS_ECDH_ECDSA_WITH_CAMELLIA_256_CBC_SHA384
TLS_ECDH_RSA_WITH_CAMELLIA_128_CBC_SHA256
TLS_ECDH_RSA_WITH_CAMELLIA_256_CBC_SHA384
TLS_RSA_WITH_CAMELLIA_128_GCM_SHA256
TLS_RSA_WITH_CAMELLIA_256_GCM_SHA384
TLS_DHE_RSA_WITH_CAMELLIA_128_GCM_SHA256
TLS_DHE_RSA_WITH_CAMELLIA_256_GCM_SHA384
TLS_ECDHE_ECDSA_WITH_CAMELLIA_128_GCM_SHA256
TLS_ECDHE_ECDSA_WITH_CAMELLIA_256_GCM_SHA384
TLS_ECDH_ECDSA_WITH_CAMELLIA_128_GCM_SHA256
TLS_ECDH_ECDSA_WITH_CAMELLIA_256_GCM_SHA384
TLS_ECDHE_RSA_WITH_CAMELLIA_128_GCM_SHA256
TLS_ECDHE_RSA_WITH_CAMELLIA_256_GCM_SHA384
TLS_ECDH_RSA_WITH_CAMELLIA_128_GCM_SHA256
TLS_ECDH_RSA_WITH_CAMELLIA_256_GCM_SHA384
TLS_PSK_WITH_CAMELLIA_128_GCM_SHA256
TLS_PSK_WITH_CAMELLIA_256_GCM_SHA384
TLS_DHE_PSK_WITH_CAMELLIA_128_GCM_SHA256
TLS_DHE_PSK_WITH_CAMELLIA_256_GCM_SHA384
TLS_RSA_PSK_WITH_CAMELLIA_128_GCM_SHA256
TLS_RSA_PSK_WITH_CAMELLIA_256_GCM_SHA384
TLS_ECJPAKE_WITH_AES_128_CCM_8

I as well testet to find the suites via ID and SSL_CTX with security level set to
0 but this of course didnt affect the internal ssl3_ciphers table

Finally my question:
Am I missing some more flags building Openssl to enable these suites? 
Or am I missing something else?
Could it be that the security level hast to be set to 0 before the compilation or something like that?

I highly appreciat any help!
Kind regards
Tamec

**Comment by mattcaswell** [comment, marked as answer]
The ciphersuites you list are not supported by OpenSSL at all for a variety of reasons:

- We don't support any single DES ciphersuites (i.e. any ciphersuite with just `_DES_` in the name)
- We don't support non-ephemeral ECDH (i.e. any ciphersuite with just `_ECDH_` and not `_ECDHE_`). Support for this used to exist in earlier versions of OpenSSL but they were removed (no one uses these and they are insecure)
- We only support GCM based ARIA ciphersuites (i.e. not CBC ones)
- We only support CBC based CAMELLIA ciphersuites (i.e. not GCM ones)
- We don't support ECJPAKE (this one doesn't seem to exist in the IANA registry at all - at least I can't see it).

I could see us supporting the CAMELLIA GCM ciphersuites at some point potentially, but it is very unlikely we would add support for any of the others.

**Reply by Tamec1** [reply]
I already thought this might be the case but didn't want to ask as i couldn't imagine openssl not supporting all ciphersuites
that have been out there so far, at least if compiled at the lowest security level.
That as well explains why I could not find any source code related to those suites.

So now im asured that we have to somehow use a proxy who decides to still use the mbed lib 
if openssl does not support the requested suite our clients like to use.

Thank you verry much for your quick reply :-)

**Comment by Tamec1** [comment]
I had to reopen this since I found out about the legacy-provider...

Just twi quick questions. 
Is it possible to use the mentioned suites using the legacy-provider?
Or is there no workaround at all since it is by purpose of OpenSSL to generally not support those suites anymore?

**Reply by t8m** [reply]
Actually, you have to have the legacy provider loaded to support any RC4, SEED or IDEA based ciphersuites.

But none of the ciphersuites you've listed above will be added by loading the legacy provider.

**Reply by t8m** [reply]
It was never the intention of the project to support all the possible ciphersuites. Especially if those are insecure or obsolete in other ways.

**Reply by mattcaswell** [reply]
The `ssl3_ciphers` table you referred to in your original question is a fixed table of ciphersuites influenced by compile time options only (notably enable-weak-ssl-ciphers). None of the ciphersuites you mentioned exist in that table. At run time OpenSSL will check which ciphers/digests etc it has available to it based on the loaded providers and will disable more ciphersuites from the table. But no amount of loading new providers will make ciphersuites not in that table available.

So to answer your question: No, loading the legacy provider will not make the ciphersuites you list available.
There is no workaround to make those ciphersuites available.

**Reply by Tamec1** [reply]
Ok thanks again for your quick replies.
I think I get it now. It also makes sense to not support such suites from an OpenSSL point of view.
For testing of e.g. older devices we simply have to use other libraries.
